### PR TITLE
(de)serialization for the VARIANT type

### PR DESCRIPTION
**Pull request by ryzhyk**

### CODE REVIEW

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 1]
```diff
@@ -387,18 +551,230 @@ where
...
+    #[error("{0}")]
+    CustomError(String),
+}
+
+impl MyError {
+    pub fn from_string(message: String) -> Box<Self> {
+        Box::new(MyError::CustomError(message))
+    }
+
+    pub fn from_strng(message: &str) -> Box<Self> {
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:567]:
```suggestion
    pub fn from_str(message: &str) -> Box<Self> {
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 2]
```diff
@@ -387,18 +551,230 @@ where
...
+macro_rules! into {
+    ($variant:ident, $type:ty) => {
+        impl TryFrom<Variant> for $type {
+            type Error = Box<dyn Error>;
+
+            fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                match value {
+                    Variant::$variant(x) => Ok(x),
+                    _ => Err(MyError::from_strng(concat!(
+                        "Not an ",
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:581]:
```suggestion
                        "not an ",
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 3]
```diff
@@ -387,18 +551,230 @@ where
...
+into_numeric!(i64, i64);
+into_numeric!(F32, f);
+into_numeric!(F64, d);
+
+impl TryFrom<Variant> for Decimal {
+    type Error = Box<dyn Error>;
+
+    fn try_from(value: Variant) -> Result<Self, Self::Error> {
+        match value {
+            Variant::TinyInt(x) => Decimal::from_i8(x).ok_or(MyError::from_strng("Out of range")),
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:666]:
```suggestion
            Variant::TinyInt(x) => Decimal::from_i8(x).ok_or(MyError::from_strng("out of range")),
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 4]
```diff
@@ -387,18 +551,230 @@ where
...
+                fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                    match value {
+                        Variant::TinyInt(x) => Ok([< cast_to_ $type_name _i8>](x)),
+                        Variant::SmallInt(x) => Ok([< cast_to_ $type_name _i16>](x)),
+                        Variant::Int(x) => Ok([< cast_to_ $type_name _i32 >](x)),
+                        Variant::BigInt(x) => Ok([< cast_to_ $type_name _i64 >](x)),
+                        Variant::Real(x) => Ok([< cast_to_ $type_name _f >](x)),
+                        Variant::Double(x) => Ok([< cast_to_ $type_name _d >](x)) ,
+                        Variant::Decimal(x) => Ok([< cast_to_ $type_name _decimal >](x)),
+                        _ => Err(MyError::from_strng(concat!("Not an ", stringify!($variant)))),
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:629]:
```suggestion
                        _ => Err(MyError::from_strng(concat!("not an ", stringify!($variant)))),
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 5]
```diff
@@ -387,18 +551,230 @@ where
...
+
+macro_rules! into_numeric {
+    ($type:ty, $type_name: ident) => {
+        impl TryFrom<Variant> for $type {
+            type Error = Box<dyn Error>;
+
+            ::paste::paste! {
+                fn try_from(value: Variant) -> Result<Self, Self::Error> {
+                    match value {
+                        Variant::TinyInt(x) => Ok([< cast_to_ $type_name _i8>](x)),
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:622]: many of these can panic, can't they? I think we should use non-panicking versions here

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 6]
```diff
@@ -387,18 +551,230 @@ where
...
+impl<T> TryFrom<Variant> for Vec<T>
+where
+    T: TryFrom<Variant, Error = Box<dyn Error>>,
+{
+    type Error = Box<dyn Error>;
+
+    fn try_from(value: Variant) -> Result<Self, Self::Error> {
+        match value {
+            Variant::Array(a) => {
+                let mut result = vec![];
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:709]:
```suggestion
                let mut result = Vec::with_capacity(a.len());
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 7]
```diff
@@ -387,18 +551,230 @@ where
...
+                for e in a {
+                    let converted = T::try_from(e)?;
+                    result.push(converted);
+                }
+                Ok(result)
+            }
+            _ => Err(Box::new(MyError::CustomError("Not an array".to_string()))),
```
Comment (ryzhyk) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:716]:
```suggestion
            _ => Err(Box::new(MyError::CustomError("not an array".to_string()))),
```

**Comment by gz** [comment]
Do we believe we can support VARIANT in tables for all formats? Otherwise it would be bad UX to have certain types only work with certain formats.

**Comment by mihaibudiu** [comment]
what is a "format"?

**Comment by ryzhyk** [comment]
> Do we believe we can support VARIANT in tables for all formats? Otherwise it would be bad UX to have certain types only work with certain formats.

Good question. I think it works for any format that has a notion of dynamically typed objects, which most formats do.

**Comment by ryzhyk** [comment]
But the main use case for this is semi-structured data, so mostly things like JSON, where you want to ingest data without a strongly typed schema.

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 8]
```diff
@@ -387,18 +551,240 @@ where
...
+
+    pub fn from_strng(message: &str) -> Box<Self> {
+        Box::new(ConversionError::CustomError(message.to_string()))
+    }
+}
+
+macro_rules! into {
+    ($variant:ident, $type:ty) => {
+        impl TryFrom<Variant> for $type {
+            type Error = Box<dyn Error>;
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:575]: probably shouldn't have to Box the error?

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 9]
```diff
@@ -387,18 +551,240 @@ where
...
+                    _ => Err(ConversionError::from_strng(concat!(
+                        "not an ",
+                        stringify!($variant)
+                    ))),
+                }
+            }
+        }
+
+        impl TryFrom<Variant> for Option<$type> {
+            type Error = Box<dyn Error>;
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:589]: same here

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs (hunk 8 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:575]: I spent one hour trying to make the errors work. Can you show me an example?

Comment (gz) [reply sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:575]: I can try to fix it tomorrow

File: crates/pipeline-manager/src/integration_test.rs [hunk 10]
```diff
@@ -1300,16 +1300,16 @@ async fn parse_datetime() {
     let response = config
         .post_json(
             "/v0/pipelines/test/ingress/t1?format=json&update_format=raw",
-            r#"{"t":"13:22:00","ts": "2021-05-20 12:12:33","d": "2021-05-20"}
-            {"t":" 11:12:33.483221092 ","ts": " 2024-02-25 12:12:33 ","d": " 2024-02-25 "}"#
+            r#"{"t":"13:22:00","ts": "2021-05-20T12:12:33Z","d": "2021-05-20"}
```
Comment (mihaibudiu) [inline crates/pipeline-manager/src/integration_test.rs:1303]: I assume these changes are unrelated to variant

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 11]
```diff
@@ -146,29 +147,237 @@ pub enum Variant {
...
+            fn visit_none<E>(self) -> Result<Variant, E> {
+                Ok(Variant::VariantNull)
+            }
+
+            #[inline]
+            fn visit_some<D>(self, deserializer: D) -> Result<Variant, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                Deserialize::deserialize(deserializer)
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:207]: I don't understand what happens here

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 12]
```diff
@@ -146,29 +147,237 @@ pub enum Variant {
...
+                Ok(Variant::VariantNull)
+            }
+
+            #[inline]
+            fn visit_seq<V>(self, mut visitor: V) -> Result<Variant, V::Error>
+            where
+                V: SeqAccess<'de>,
+            {
+                let mut vec = Vec::new();
+
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:221]: why so many empty lines

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 13]
```diff
@@ -146,29 +147,237 @@ pub enum Variant {
...
+                Ok(Variant::Array(vec))
+            }
+
+            #[inline]
+            fn visit_map<V>(self, mut visitor: V) -> Result<Variant, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                match visitor.next_key_seed(KeyClassifier)? {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:234]: I don't understand what this

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 14]
```diff
@@ -146,29 +147,237 @@ pub enum Variant {
...
+        struct Visitor;
+
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = NumberFromString;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string containing a decimal number")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<NumberFromString, E>
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:325]: is this expected to consume the whole string?
because otherwise it's ambiguous

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 15]
```diff
@@ -416,3 +744,267 @@ pub fn typeofN(value: Option<Variant>) -> String {
...
+            r#"{
+                "b": true,
+                "i": 12345,
+                "f": 0.00123,
+                "d": 123.45,
+                "s": "foo\nbar",
+                "n": null,
+                "nested": {
+                    "arr": [1, "foo", null],
+                    "ts": "2024-12-19T16:39:57+00:00",
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:997]: json_unparse does not do this for timestamps, perhaps it could

**Review by mihaibudiu — APPROVED** [review]

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs (hunk 15 above)
Comment (ryzhyk) [reply sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:997]: it should do it now

File: crates/pipeline-manager/src/integration_test.rs (hunk 10 above)
Comment (ryzhyk) [reply crates/pipeline-manager/src/integration_test.rs:1303]: yes, this is related to the new default encoding for timestamps

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs (hunk 15 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:997]: dates and times too?

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs (hunk 14 above)
Comment (ryzhyk) [reply sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:325]: yes, it should consume the entire string that represents the decimal number.

**Comment by ryzhyk** [comment]
@gz , @mihaibudiu , please have a look at the docs.

File: docs/formats/csv.md [hunk 16]
```diff
@@ -48,6 +48,7 @@ interpreted as an empty string instead of `NULL`.
 | DATE                                    | `2024-02-25`                                    |
 | BIGINT ARRAY                            | `[1, 2]`                                        |
 | VARCHAR ARRAY ARRAY                     | `[[ 'abc', '123'], ['c', 'sql']]`               |
+| VARIANT                                 | `{""foo"": ""bar""}`                            |
```
Comment (mihaibudiu) [inline docs/formats/csv.md:51]: what happens to commas?

File: docs/sql/json.md [hunk 17]
```diff
@@ -7,28 +7,20 @@ Any such value holds at runtime two pieces of information:
...
-
-
 Values of `VARIANT` type can be created by casting any other value to
-a `VARIANT`: e.g.  `SELECT CAST(x AS VARIANT)`.  Conversely, values of
-type `VARIANT` can be cast to any other data type `SELECT CAST(variant
-AS INT)`.  A cast of a value of type `VARIANT` to target type T will
-compare the runtime type with T.  If the types are identical or there
-is a natural conversion from the runtime type to T, the original value
+a `VARIANT`: e.g. `SELECT CAST(x AS VARIANT)`.  Conversely, values of
+type `VARIANT` can be cast to any other data type`SELECT CAST(variant AS INT)`.
```
Comment (mihaibudiu) [inline docs/sql/json.md:12]: space is missing. Maybe you want an e.g.

File: docs/sql/json.md [hunk 18]
```diff
@@ -7,28 +7,20 @@ Any such value holds at runtime two pieces of information:
...
+A cast of a value of type `VARIANT` to target type `T` will
+compare the runtime type with `T`.  If the types are identical or there
```
Comment (mihaibudiu) [inline docs/sql/json.md:14]: I don't think we convert to strings yet.
could be done, though

File: docs/sql/json.md [hunk 19]
```diff
@@ -55,25 +47,13 @@ the JSON `null` value.  An important difference is that two `VARIANT`
   is subject to the capitalization rules of the SQL dialect, so for correct
   operation the field may need to be quoted: `variant."field"`
 
-The runtime types do not match exactly the compile-time types:
```
Comment (mihaibudiu) [inline docs/sql/json.md:58]: I think this is important and should not be deleted

File: docs/sql/json.md [hunk 20]
```diff
@@ -100,20 +80,89 @@ SELECT CAST(
...
+
+Input events can use any [supported data format](/docs/formats/).  For instance, when
+ingesting a [JSON stream](/docs/formats/json), a valid input record could look like this
+(note the use of escaping in the `json` field):
+
+```json
+{"id": 123, "json": "{\"foo\": \"bar\"}"}
+```
+
+The second approach eliminates the need for the intermediate view by
```
Comment (mihaibudiu) [inline docs/sql/json.md:122]: it doesn't have to be a view, in general it's just a select that calls parse_json.

File: python/tests/test_variant.py [hunk 21]
```diff
@@ -10,36 +10,92 @@
...
-CREATE LOCAL VIEW tmp AS SELECT PARSE_JSON(json) AS json FROM json;
-
-CREATE VIEW average AS SELECT
+CREATE MATERIALIZED VIEW average_view AS SELECT
 CAST(json['name'] AS VARCHAR) as name,
 ((CAST(json['scores'][1] AS DECIMAL(8, 2)) + CAST(json['scores'][2] AS DECIMAL(8, 2))) / 2) as average
-FROM tmp;
+FROM json_view;
+
+-- Ingest JSON as variant; extract strongly types columns from it.
```
Comment (mihaibudiu) [inline python/tests/test_variant.py:23]: typed

File: docs/sql/json.md [hunk 22]
```diff
@@ -7,28 +7,20 @@ Any such value holds at runtime two pieces of information:
...
+is a natural conversion from the runtime type to `T`, the original value
 is returned.  Otherwise the `CAST` returns `NULL`.
 
 ::: warning
 
 Remember that the `DECIMAL` type specified without precision is the
 same as `DECIMAL(0)`, with no digits after the decimal point.  When
 you cast a `VARIANT` value to `DECIMAL` you should specify a precision
-and scale good enough for the values that you expect in the data.
+and scale large enough for the values that you expect in the data.
```
Comment (mihaibudiu) [inline docs/sql/json.md:23]: we don't support user-defined types yet, maybe this deserves a mention

**Review by mihaibudiu — APPROVED** [review]

File: docs/formats/parquet.md [hunk 23]
```diff
@@ -28,6 +28,7 @@ and [Arrow types](https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.htm
 | `ARRAY`                                    | `DataType::LargeList`                                                              |
 | `STRUCT`                                   | `DataType::Struct`                                                                 |
 | `MAP`                                      | `DataType::Dictionary`                                                             |
+| `VARIANT`                                  | `LargeUtf8`                                                                        |
```
Comment (gz) [inline docs/formats/parquet.md:31]:
```suggestion
| `VARIANT`                                  | `LargeUtf8`  (serialized as JSON)                                         |
```

**Review by gz — APPROVED** [review]

File: docs/sql/json.md (hunk 19 above)
Comment (ryzhyk) [reply docs/sql/json.md:58]: It sounds like an internal implementation detail the way it's documented. If it's important, let's explain what's actually important to the user.

Comment (mihaibudiu) [reply docs/sql/json.md:58]: It's important because of two reasons : the types can be used using type of in a case statement to achieve the effect of rust match. Second, it tells you which casts will work

Comment (mihaibudiu) [reply docs/sql/json.md:58]: Will add an example using case

File: docs/sql/json.md (hunk 18 above)
Comment (ryzhyk) [reply docs/sql/json.md:14]: Not sure what you mean. Which conversion is missing?

File: docs/sql/json.md (hunk 19 above)
Comment (ryzhyk) [reply docs/sql/json.md:58]: Let's do a separate PR.

File: docs/sql/json.md (hunk 18 above)
Comment (mihaibudiu) [reply docs/sql/json.md:14]: Anything to string

Comment (ryzhyk) [reply docs/sql/json.md:14]: Even if the variant contains string? What can you cast it to then?

Comment (mihaibudiu) [reply docs/sql/json.md:14]: If it does not contain string

**Comment by mihaibudiu** [comment]
The ci failed because some real numbers are not exactly as they used to be

File: docs/sql/json.md [hunk 24]
```diff
@@ -100,12 +81,9 @@ SELECT CAST(
       ] AS VARIANT)
 ```
 
-
 ### `UNPARSE_JSON`
 
-`PARSE_JSON` converts a `VARIANT` value to a `VARCHAR`.  The
-assumption is that the `VARIANT` was has the structure as produced by
-the `PARSE_JSON` function:
+`UNPARSE_JSON` converts a `VARIANT` value to a `VARCHAR`:
```
Comment (lalithsuresh) [inline docs/sql/json.md]: Shouldn't this be to_json() ?
